Create "chromeConfigControls" for plugins to add chrome config sections

trevan opened the pull request:
The timepicker config and navrbar section has been moved over to this registry so you can get an idea of what it would look like.  I also removed the chorome/context.js since that is now inside of timepicker/toggle.js.

I did not convert the app switcher since that appears to be going away in the redesign.

I have not yet added unittests for the new "kbnChromeConfigControls" directive but all the other tests should pass.

This closes #6136

elasticsearch-release commented:
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run; then  say 'jenkins, test it'.

epixa commented:
This needs to be rebased on master or have master merged into it.

trevan commented:
I merged master into it.

epixa commented:
jenkins, test it

spalger commented on src/ui/public/chrome/chrome.html:
```diff
@@ -56,6 +56,8 @@
     <!-- /Full navbar -->
   </nav>
 
+  <div kbn-chrome-config-controls>
```
nitpick, can we just make this a custom element, `<kbn-chrome-config-controls>`

spalger commented on src/ui/public/timepicker/toggle.js:
```diff
@@ -1,16 +1,43 @@
-// logic here
+chromeConfigControlsRegistry.register(function (timefilter, globalState) {
+  let pickerTemplate = new ConfigTemplate({
+    filter: require('ui/timepicker/config/filter.html'),
```
These templates should be imported, rather than required

spalger commented:
This is looking great @trevan

trevan commented:
@spalger, I updated per your comments.  I also added tests for the new directive.  And I modified the append nav directive to merge the items from the nav registry and the config registry so you can control their order.

epixa commented:
jenkins, test it

spalger commented:
This LGTM. Opened #6224 which merged this with master. This is ready for another set of eyes. Working on tracking down a candidate.

elasticsearch-release commented:
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run; then  say 'jenkins, test it'.

trevan commented:
@spalger, I noticed that the autorefresh functionality was broken with these changes so I've merged in master and fixed the issue.

epixa commented:
jenkins, test it

epixa commented:
jenkins, test it

spalger commented on src/ui/public/chrome/directives/append_nav_controls.js:
```diff
@@ -17,16 +13,42 @@ export default function (chrome, internals) {
+          }),
+
+          ...configs.map(function (config) {
+            const configHtml = `<render-directive definition="configs['${config.name}'].navbar">
```
Not sure how I missed this the first time, but we should probably take a tad-bit more care with how we build `configHtml`

trevan replied:
What do you mean?

spalger replied:
Well, I guess we have to trust the template value, but I would feel better if we changed the injection of `config.name` to something like:

```
`<render-directive definition="configs.byName['${JSON.stringify(config.name)}'].navbar">
```

trevan replied:
The JSON.stringify didn't work since it add double quotes which messed up the html parsing.  I instead add a regex piece in the registry that forces the names to only allow a-z, A-Z, or 0-9.

spalger commented on src/ui/public/courier/courier.js:
```diff
@@ -131,9 +131,9 @@ uiModules.get('kibana/courier')
-    $rootScope.$watchCollection('timefilter.refreshInterval', function () {
-      var refreshValue = _.get($rootScope, 'timefilter.refreshInterval.value');
-      var refreshPause = _.get($rootScope, 'timefilter.refreshInterval.pause');
+    $rootScope.$watchCollection('$$timefilter.refreshInterval', function () {
```
Yeah, we can't use the `$$timefilter` here. The `$$` prefix is explicitly designed to prevent it's use. Maybe we can inject the timefilter and use a getter function in the `watchCollection`

spalger replied:
https://gist.github.com/anonymous/a437532d06f06242ae3a

trevan replied:
That gist worked for me so I committed it.

spalger commented:
Awesome, thanks for the updates @trevan!

spalger commented:
Assigned to @lukasolson to get a second set of eyes on the changes

lukasolson commented:
Looks like we need master merged again... :)

trevan commented:
@lukasolson, I merged master

spalger commented:
Alright, so with the updated coming in [feature/redesign](https://github.com/elastic/kibana/tree/feature/design) I think we need to put this on hold momentarily. Sorry @trevan.

Once we get feature/design merged we can find a way to get this into there. It shouldn't be too hard, but since the chrome is now on the left of the screen and the taskbar is implemented by each application, rather than injected by the chrome, we will need to think about how to implement this there.

trevan commented:
@spalger, ok.  I still have a use for a control next to the timepicker and I don't want it to just be a static string either.

panda01 commented:
This is no longer blocked, though maybe you were able to achieve what you wanted without this, either way @trevan, would you like to continue this?

panda01 commented:
@trevan i understand you're busy, but due to a lack of feedback, I'm going to close this PR. If you would like to submit another one under the same branch it is always welcome!
